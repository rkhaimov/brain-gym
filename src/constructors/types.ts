import { Either } from '../either';

export type TypeConstructor<TLeft, TRight> = (
  value: unknown
) => Either<TLeft, TRight>;
