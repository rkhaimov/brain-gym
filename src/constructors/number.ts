import { TypeConstructor } from './types';
import { Either } from '../either';

export const number = (): TypeConstructor<string, number> => (value) => {
  if (typeof value === 'number') {
    return Either.right(value);
  }

  return Either.left('is not a number');
};

