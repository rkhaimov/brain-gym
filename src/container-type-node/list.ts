import { ToType, TypeNode } from '../core';
import { fromErrorMessage, prependErrorPathWith, validate } from '../validate';
import {
  ContainerTypeNode,
  createContainerTypeNode,
} from './createContainerTypeNode';

export const list = <TChildren extends TypeNode>(
  tn: TChildren
): ContainerTypeNode<ToType<TChildren>[], TChildren> =>
  createContainerTypeNode({
    validate: (value) => {
      if (Array.isArray(value)) {
        return value.flatMap((element, index) =>
          validate(tn, element).map(prependErrorPathWith(index))
        );
      }

      return [fromErrorMessage('Not a list')];
    },
    defaults: () => [],
    children: () => tn,
  });
