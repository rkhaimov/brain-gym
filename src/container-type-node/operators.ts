// TODO: Should generalize to all structures (tuples, trees, graphs)
import { ToType, TypeNode, TypeNodeOperator } from '../core';
import { ContainerTypeNode } from './createContainerTypeNode';
import { andValidate, defaultsTo } from '../operators';
import { fromErrorMessage } from '../validate';

export const nonEmptyList =
  <TChildren extends TypeNode>(): TypeNodeOperator<
    ContainerTypeNode<ToType<TChildren>[], TChildren>
  > =>
  (tn) =>
    tn.wrap(
      defaultsTo(() => [tn.children().defaults() as ToType<TChildren>]),
      andValidate((value) => {
        if (value.length === 0) {
          return [fromErrorMessage('list is empty')];
        }

        return [];
      })
    );
