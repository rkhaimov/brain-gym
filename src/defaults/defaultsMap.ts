import { ToType, TypeNode, TypeNodeOperator } from '../core';

export const defaultsMap =
  <TTypeNode extends TypeNode>(
    transform: (defaults: ToType<TTypeNode>, tn: TTypeNode) => ToType<TTypeNode>
  ): TypeNodeOperator<TTypeNode> =>
  (tn) =>
    tn.__clone({
      defaults: () => transform(tn.defaults() as ToType<TTypeNode>, tn),
    });
