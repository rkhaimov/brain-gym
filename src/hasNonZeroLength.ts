import { TypeNode, TypeNodeOperator } from './core';
import { fromKindAndPayload } from './validate';
import { validateConcatMap } from './validateConcatMap';

declare module 'validation-messages' {
  interface ValidationMessages {
    hasNonZeroLength: void;
  }
}

export const hasNonZeroLength =
  <
    TType extends { length: number },
    TTypeNode extends TypeNode<TType>
  >(): TypeNodeOperator<TTypeNode> =>
  (tn) =>
    tn.wrap(
      validateConcatMap((value) => {
        if (value.length === 0) {
          return [fromKindAndPayload('hasNonZeroLength')];
        }

        return [];
      })
    );
