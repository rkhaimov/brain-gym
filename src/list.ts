// Either empty or not
type List<T> = Either<void, [T, List<T>]>;

declare function createEmptyList<T>(): List<T>;

declare function followedBy<T>(element: T, origin: List<T>): List<T>;

