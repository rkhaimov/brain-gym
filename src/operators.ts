import { ToType, TypeNode, TypeNodeOperator } from './core';

export const defaultsTo =
  <TTypeNode extends TypeNode>(
    defaults: () => ToType<TTypeNode>
  ): TypeNodeOperator<TTypeNode> =>
  (tn) =>
    tn.clone({ validate: tn.validate, defaults });

export const andValidate =
  <TTypeNode extends TypeNode>(
    validate: TTypeNode['validate']
  ): TypeNodeOperator<TTypeNode> =>
  (tn) =>
    tn.clone({
      validate: (value) => [...tn.validate(value), ...validate(value)],
      defaults: tn.defaults,
    });
