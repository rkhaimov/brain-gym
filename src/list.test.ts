import { list } from './container-type-node/list';
import { defaultsMap } from './defaultsMap';
import { hasNonZeroLength } from './hasNonZeroLength';
import { translateTo } from './translateTo';
import { biggerOrEqThan } from './type-node/operators/biggerOrEqThan';
import { number } from './type-node/types/number';
import { validate } from './validate';

describe('list', () => {
  describe('Validates and generates defaults', () => {
    it('validates with static checking', () => {
      const tn = list(number());

      expectNoErrors(validate(tn, []));

      expectNoErrors(validate(tn, [42]));

      // @ts-expect-error // Static check
      expect(validate(tn, '42')).toMatchSnapshot();

      expect(validate(tn, ['42', 42, '42'])).toMatchSnapshot();
    });

    it('generates defaults for composite types', () => {
      const tn = list(number().wrap(biggerOrEqThan(10))).wrap(
        hasNonZeroLength(),
        defaultsMap((_, tn) => [tn.children().defaults()]),
        translateTo({ biggerOrEqThan: () => 'Number is too small' })
      );

      expect(validate(tn, [])).toMatchSnapshot();
      expectNoErrors(validate(tn, tn.defaults()));
      expect(validate(tn, [2])).toMatchSnapshot();
    });
  });
});

function expectNoErrors(errors: unknown[]) {
  expect(errors.length).toBe(0);
}
