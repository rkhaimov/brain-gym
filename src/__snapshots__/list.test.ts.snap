// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list Validates and generates defaults generates defaults for composite types 1`] = `
Array [
  Object {
    "kind": "hasNonZeroLength",
    "message": "hasNonZeroLength: no params",
    "params": undefined,
    "paths": Array [],
  },
]
`;

exports[`list Validates and generates defaults generates defaults for composite types 2`] = `
Array [
  Object {
    "kind": "biggerOrEqThan",
    "message": "Number is too small",
    "params": 10,
    "paths": Array [
      0,
    ],
  },
]
`;

exports[`list Validates and generates defaults validates with static checking 1`] = `
Array [
  Object {
    "kind": "unknown",
    "message": "Not a list",
    "params": undefined,
    "paths": Array [],
  },
]
`;

exports[`list Validates and generates defaults validates with static checking 2`] = `
Array [
  Object {
    "kind": "number",
    "message": "number: no params",
    "params": undefined,
    "paths": Array [
      0,
    ],
  },
  Object {
    "kind": "number",
    "message": "number: no params",
    "params": undefined,
    "paths": Array [
      2,
    ],
  },
]
`;
